text/0018-bee-bundle/0018-bee-bundle.md

```diff
@@ -0,0 +1,281 @@
+
+### Bundle
+
+As bundles are final, they shouldn't be modifiable outside of the scope of the bundle module.
+
+A bundle can simply be represented as an array of transactions:
+
+```rust
+struct Bundle {
+    transactions: Vec<Transaction>
+}
+```
+
+And its implementation should only allow to retrieve transactions:
+
+```rust
+impl Bundle {    
+    pub fn transactions(&self) -> &Vec<Transaction> {
+        &self.transactions
+    }
+}
+```
+
+### BundleBuilder
+
+The `BundleBuilder` offers a simple and convenient way to build bundles:
+
+```rust
+pub struct BundleBuilder {
+    transactions: Vec<TransactionBuilder>
+}
+```
+
+```rust
+impl BundleBuilder {    
+    pub fn hash(&self) {
+      unimplemented!()
+    }
+
+    pub fn finalise(&self) {
+      unimplemented!()
+    }
+
+    pub fn sign(&self) {
+      unimplemented!()
+    }
+
+    pub fn pow(&self) {
+      unimplemented!()
+    }
+
+    pub fn validate(&self) {
+      unimplemented!()
+    }
+
+    pub fn build(&self) {
+      unimplemented!()
+    }
+}
+```
+
+*We do not list parameters and/or return values as they are implementation details.*
+
+## Algorithms
+
+In this section, we describe the algorithms needed to build a `Bundle`. The lifecycle of a `BundleBuilder` depends on if it's being used in client side or server side:
+- client side: `finalise` -> [`sign` ->] -> [`pow` ->] -> `validate` -> `build`
+- server side: `addTransaction`/`addTransactionBuilder` -> `validate` -> `build`
+
+*`sign` is optional because data transactions don't have to be signed. `pow` is optional because one can use remote pow instead.*
+
+### Hash
+
+*Client side and server side operation.*
+
+A bundle hash ties different transactions together. By having this common hash in their `bundle` field, it makes it clear that these transactions should be processed as a whole.
+
+The hash of a bundle is derived from the bundle essence of each of its transactions. The bundle essence of a transaction is composed of the following fields.
+
+| Name          | Size      |
+| ------------- | --------- |
+| address       | 243 trits |
+| value         | 81 trits  |
+| obsolete_tag  | 81 trits  |
+| timestamp     | 27 trits  |
+| current_index | 27 trits  |
+| last_index    | 27 trits  |
+
+The bundle hash is generated with a sponge by iterating through the bundle, from `0` to `last_index`, absorbing the bundle essence of each transaction and eventually squeezing the bundle hash from the sponge.
+
+Pseudocode:
+
+```
+hash(bundle)
+| sponge = Sponge(HASH_FUNCTION)
+|
+| for transaction in bundle
+| | sponge.absorb(transaction.essence())
+|
+| return sponge.squeeze()
+```
+
+*In the current mainnet, the hash function of the sponge used to generate bundle hashes is Kerl.*
```

Comment:
Not sure if we should write "in the current mainnet". I mean this bundle proposal is directly for the mainnet. We shouldn't need to refer it.

Reply:
You're right